File: binarysearch/32_search_in_rotated_array/main.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
    public:
        int search(vector<int>& nums, int target) {
            int start = 0, end = nums.size() - 1;
            // int ans = -1;
            while (start <= end) {
                long long mid = start + (end - start)/2;
    
                if (nums[mid] == target) return mid;
    
                if (nums[mid] >= nums[start]) {
                    // Check the left sorted part
                    if (target >= nums[start] && nums[mid] > target) {
                        end = mid - 1;
                    } else {
                        start = mid + 1;
                    }
                }
                else {
                    // Check the right sorted part
                    if (target <= nums[end] && nums[mid] < target) {
                        start = mid + 1;
                    } else {
                        end = mid - 1;
                    }
                }
            }
    
            return -1;
        }
    };

File: binarysearch/32_search_in_rotated_array/main.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        start, end = 0, len(nums)-1

        while start <= end:
            mid = start + (end - start)//2

            if nums[mid] == target:
                return mid

            # Check the left sorted part
            if nums[mid] >= nums[start]:
                if target >= nums[start] and nums[mid] > target:
                    end = mid - 1
                else:
                    start = mid + 1

            # Check the right sorted part
            else:
                if target <= nums[end] and nums[mid] < target:
                    start = mid + 1
                else:
                    end = mid - 1

        return -1